File: javascript/src/metrics.ts
/** Routing evaluation metrics: accuracy, macro-F1, confusion matrix. */

import { PRIMARY_ACTIONS } from "./schema.ts";

export function confusionMatrix(
  yTrue: readonly string[],
  yPred: readonly string[],
  labels?: readonly string[],
): number[][] {
  const resolved = labels ? [...labels] : [...PRIMARY_ACTIONS];
  const index = new Map(resolved.map((label, i) => [label, i]));
  const size = resolved.length;
  const matrix = Array.from({ length: size }, () => Array.from({ length: size }, () => 0));
  if (yTrue.length !== yPred.length) {
    throw new Error("y_true and y_pred must have the same length");
  }
  for (let i = 0; i < yTrue.length; i += 1) {
    const truth = index.get(yTrue[i]);
    const pred = index.get(yPred[i]);
    if (truth !== undefined && pred !== undefined) {
      matrix[truth][pred] += 1;
    }
  }
  return matrix;
}

export function accuracyScore(yTrue: readonly string[], yPred: readonly string[]): number {
  if (yTrue.length !== yPred.length) {
    throw new Error("y_true and y_pred must have the same length");
  }
  if (!yTrue.length) {
    return 0.0;
  }
  let hits = 0;
  for (let i = 0; i < yTrue.length; i += 1) {
    if (yTrue[i] === yPred[i]) {
      hits += 1;
    }
  }
  return hits / yTrue.length;
}

export function macroF1Score(
  yTrue: readonly string[],
  yPred: readonly string[],
  labels?: readonly string[],
): number {
  const resolved = labels ? [...labels] : [...PRIMARY_ACTIONS];
  if (!resolved.length) {
    return 0.0;
  }
  const matrix = confusionMatrix(yTrue, yPred, resolved);
  const scores: number[] = [];
  const n = resolved.length;
  for (let i = 0; i < n; i += 1) {
    const truePositive = matrix[i][i];
    let falsePositive = 0;
    let falseNegative = 0;
    for (let row = 0; row < n; row += 1) {
      if (row !== i) {
        falsePositive += matrix[row][i];
      }
    }
    for (let col = 0; col < n; col += 1) {
      if (col !== i) {
        falseNegative += matrix[i][col];
      }
    }
    const precision = truePositive + falsePositive ? truePositive / (truePositive + falsePositive) : 0.0;
    const recall = truePositive + falseNegative ? truePositive / (truePositive + falseNegative) : 0.0;
    if (precision + recall === 0.0) {
      scores.push(0.0);
    } else {
      scores.push((2.0 * precision * recall) / (precision + recall));
    }
  }
  return scores.reduce((sum, value) => sum + value, 0) / scores.length;
}

export function routingMetrics(
  yTrue: readonly string[],
  yPred: readonly string[],
  labels?: readonly string[],
): {
  accuracy: number;
  macro_f1: number;
  labels: string[];
  confusion_matrix: number[][];
} {
  const resolved = labels ? [...labels] : [...PRIMARY_ACTIONS];
  return {
    accuracy: accuracyScore(yTrue, yPred),
    macro_f1: macroF1Score(yTrue, yPred, resolved),
    labels: resolved,
    confusion_matrix: confusionMatrix(yTrue, yPred, resolved),
  };
}

export function formatConfusion(cm: readonly (readonly number[])[], labels: readonly string[]): string {
  const header = "pred→\\true↓ | " + labels.map((label) => label.padStart(14, " ")).join(" | ");
  const lines = [header, "-".repeat(header.length)];
  for (let i = 0; i < labels.length; i += 1) {
    const row = labels.map((_, j) => String(cm[i][j]).padStart(14, " ")).join(" | ");
    lines.push(`${labels[i].padStart(14, " ")} | ${row}`);
  }
  return lines.join("\n");
}

File: src/fta_router/metrics.py
"""Routing evaluation metrics: accuracy, macro-F1, confusion matrix.

Implemented in the standard library so a core install does not pull
NumPy or scikit-learn.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from .schema import PRIMARY_ACTIONS


def confusion_matrix(
    y_true: Sequence[str],
    y_pred: Sequence[str],
    labels: Sequence[str] | None = None,
) -> list[list[int]]:
    """Return a rows-true / cols-pred confusion matrix over *labels*."""
    labels = list(labels) if labels is not None else list(PRIMARY_ACTIONS)
    index = {label: i for i, label in enumerate(labels)}
    size = len(labels)
    matrix = [[0] * size for _ in range(size)]
    if len(y_true) != len(y_pred):
        raise ValueError("y_true and y_pred must have the same length")
    for truth, pred in zip(y_true, y_pred, strict=True):
        if truth in index and pred in index:
            matrix[index[truth]][index[pred]] += 1
    return matrix


def accuracy_score(y_true: Sequence[str], y_pred: Sequence[str]) -> float:
    """Fraction of exact label matches. Empty input is 0.0."""
    if len(y_true) != len(y_pred):
        raise ValueError("y_true and y_pred must have the same length")
    if not y_true:
        return 0.0
    hits = sum(truth == pred for truth, pred in zip(y_true, y_pred, strict=True))
    return hits / len(y_true)


def macro_f1_score(
    y_true: Sequence[str],
    y_pred: Sequence[str],
    labels: Sequence[str] | None = None,
) -> float:
    """Unweighted mean of per-class F1 (zero_division=0)."""
    labels = list(labels) if labels is not None else list(PRIMARY_ACTIONS)
    if not labels:
        return 0.0
    matrix = confusion_matrix(y_true, y_pred, labels=labels)
    scores: list[float] = []
    n = len(labels)
    for i in range(n):
        true_positive = matrix[i][i]
        false_positive = sum(matrix[row][i] for row in range(n) if row != i)
        false_negative = sum(matrix[i][col] for col in range(n) if col != i)
        precision = (
            true_positive / (true_positive + false_positive)
            if (true_positive + false_positive)
            else 0.0
        )
        recall = (
            true_positive / (true_positive + false_negative)
            if (true_positive + false_negative)
            else 0.0
        )
        if precision + recall == 0.0:
            scores.append(0.0)
        else:
            scores.append(2.0 * precision * recall / (precision + recall))
    return sum(scores) / len(scores)


def routing_metrics(
    y_true: Sequence[str],
    y_pred: Sequence[str],
    labels: Sequence[str] | None = None,
) -> dict[str, Any]:
    """Compute accuracy, macro-F1, and confusion matrix for ``primary_action``."""
    labels = list(labels) if labels is not None else list(PRIMARY_ACTIONS)
    matrix = confusion_matrix(y_true, y_pred, labels=labels)
    return {
        "accuracy": float(accuracy_score(y_true, y_pred)),
        "macro_f1": float(macro_f1_score(y_true, y_pred, labels=labels)),
        "labels": labels,
        "confusion_matrix": matrix,
    }


def format_confusion(cm: Sequence[Sequence[int]], labels: Sequence[str]) -> str:
    """Pretty-print a confusion matrix (rows = gold, columns = predicted)."""
    header = "pred→\\true↓ | " + " | ".join(f"{label:>14}" for label in labels)
    lines = [header, "-" * len(header)]
    for i, label in enumerate(labels):
        row = " | ".join(f"{cm[i][j]:14d}" for j in range(len(labels)))
        lines.append(f"{label:>14} | {row}")
    return "\n".join(lines)
